content/blog/mozilla/2018-07-03.md

```diff
@@ -0,0 +1,217 @@
+needed to implement GC barriers to make sure that the stored value would not be
+marked as unused during tracing. Our garbage collector does incremental marking
+thus it needs pre-barriers; since it's also generational, it needs
+post-barriers. There is good literature explaining why these barriers are
+needed and what they do, so I will not expand too much on the topic.
+
+Here's an example of usage at the time being (spoiler alert, all these things
+might change, including the text format representation, type names, APIs,
+everything):
+
+```lisp
+(module
+    (func $alert (import "env" "alert") (param anyref))
+    (global $global_ref (mut anyref) (ref.null anyref))
+    (func (export "set_and_alert") (param $param anyref) (result anyref)
+        ;; Put the previous value of $global_ref on the virtual value stack.
+        get_global $global_ref
+        ;; Get the argument anyref value and store it in $global_ref.
+        get_local $param
+        set_global $global_ref
+        ;; Call the $alert method with the argument anyref value.
+        get_local $param
+        call $alert
+        ;; The previous value of $global_ref is still on the stack and will be
+        ;; returned.
+    )
+)
+```
+
+*Example of wasm text format using `anyref`.*
+
+```js
+(async function() {
+    let { instance } = await WebAssembly.instantiate(wasmBinary, {
+        env: {
+            alert(obj) {
+                alert(`Hello, ${obj.name}!`);
+            }
+        }
+    });
+    console.log(instance.exports.set_and_alert({
+        name: 'world',
+        secretVal: 42
+    }));
+    // alerts "Hello, world!", logs null
+
+    console.log(JSON.stringify(instance.exports.set_and_alert({
+        name: 'there'
+    })));
+    // alerts "Hello, there!", logs { name: 'world!', secretVal: 42 }
+})();
+```
+
+*Example of JavaScript using the module defined above, passing JS values and
+reading them from WebAssembly.*
+
+This is a very preliminary prototype and it might change in the next few
+months. If you feel adventurous, you can try it on Firefox Nightly by setting
+the `about:config` pref `javascript.options.wasm_gc` to `true`; be aware that
+the GC counterparts have not been entirely implemented yet, so this might
+result in out-of-memory situations. In any case, if you see something, [say
+something](https://bugzilla.mozilla.org/enter_bug.cgi?product=Core&component=Javascript%3A%20Web%20Assembly).
+
+Say you are a compiler developer, and you would like to port your langage to
+WebAssembly, and your language uses a GC. At the moment, the only way you can
+do this is by compiling your garbage collector to WebAssembly, and it would be
+backed by the wasm Module's memory. It is fine, but might not be efficient
+because it involves a lot of bookkeeping data structures that will also live
+within the wasm memory. Plus, there's already a very efficient, solidly tested,
+constantly improving garbage collector in your browser that uses all the
```

Comment:
I might summarize most of the rest of this (up to "At the moment, the reference types...") as  "What if Web Assembly gave you operations and types that let you directly reuse the JavaScript garbage collector?"

Reply:
I disagree here. While more concise, it hides all the benefits from reusing the existing GC implementations in browsers: efficiency, maintainability, and the fact it's long-lasting.